src/Microsoft.DotNet.Interactive.ExtensionLab/MsSqlKernel.cs

```diff
@@ -3,21 +3,219 @@
+                {
+                    _connectionCompleted.SetException(new Exception(connParams.ErrorMessage));
+                }
+                else
+                {
+                    _connectionCompleted.SetResult(connParams);
+                }
+            }
+        }
+
+        private async void HandleQueryCompleteAsync(object sender, QueryCompleteParams queryParams)
+        {
+            if (_queryCompletionHandler != null)
+            {
+                await _queryCompletionHandler(queryParams);
+            }
+        }
+
+        private void HandleIntellisenseReady(object sender, IntelliSenseReadyParams readyParams)
+        {
+            if (readyParams.OwnerUri.Equals(this._tempFileUri))
+            {
+                this._intellisenseReady = true;
+            }
+        }
+
+        public async Task ConnectAsync()
+        {
+            if (!_connected)
+            {
+                await _serviceClient.ConnectAsync(_tempFileUri, _connectionString);
+                await _connectionCompleted.Task;
+                _connected = true;
+            }
+        }
+
+        public async Task HandleAsync(SubmitCode command, KernelInvocationContext context)
+        {
+            if (!_connected)
+            {
+                return;
+            }
+
+            if (_queryCompletionHandler != null)
+            {
+                context.Display("Error: Another query is currently running. Please wait for that query to complete before re-running this cell.");
+                return;
+            }
+
+            var completion = new TaskCompletionSource<bool>();
+            _queryCompletionHandler = async queryParams =>
+            {
+                try
+                {
+                    foreach (var batchSummary in queryParams.BatchSummaries)
+                    {
+                        foreach (var resultSummary in batchSummary.ResultSetSummaries)
+                        {
+                            var subsetParams = new QueryExecuteSubsetParams()
+                            {
+                                OwnerUri = _tempFileUri,
+                                BatchIndex = batchSummary.Id,
+                                ResultSetIndex = resultSummary.Id,
+                                RowsStartIndex = 0,
+                                RowsCount = Convert.ToInt32(resultSummary.RowCount)
+                            };
+                            var subsetResult = await _serviceClient.ExecuteQueryExecuteSubsetAsync(subsetParams);
+
+                            if (subsetResult.Message != null)
+                            {
+                                context.Display(subsetResult.Message);
+                            }
+                            else
+                            {
+                                var tableString = GetTableStringForResult(resultSummary.ColumnInfo, subsetResult.ResultSubset.Rows);
+                                context.Display(tableString);
```

Comment:
This should return a data structure rather than assuming a specific MIME type.

Thoughts about `IEnumerable<IEnumerable<IEnumerable<(string name, object value)>>>` or similar?